### PR TITLE
debaml: native map collection coercion + value partials (Mcoerce-c, maps)

### PR DESCRIPTION
**Pull request by invakid404**
## Mcoerce-c MAPS (de-BAML epic #546)

Ports BAML's `coerce_map` partial-map behavior into native de-BAML for deterministic, non-union map targets. This is the **maps** slice of Mcoerce-c; the **lists** slice already shipped in #560. `coerceList` and its helpers are **untouched** — maps only.

### coerceMap (`internal/debaml/coerce.go`)
- **Object → map is always object-to-map:** `ObjectToMap` (score-bearing) is flagged before iterating, so a map is never a zero-score "clean" arm.
- **VALUE first, then KEY** (BAML `coerce_map.rs` order): a proven-parse-error **value** folds into `MapValueParseError` and the entry is **skipped** (partial map), keeping accepted entries in **INPUT key order** under their **ORIGINAL key strings**; a bad **key** declines the whole map.
- **Value child-classifier** (`coerceMapValueChild` / `provenMapValueError`) **reuses the LISTS `provenListItemError` whitelist verbatim** — a map value coerces on the raw child exactly like a list item — so only a **proven** BAML parse error is skipped. A value native merely *declined* (a deferred Mcoerce-d success, e.g. `map<string,string>` with a **number** → `JsonToString`) declines the **whole map** rather than skip an entry BAML would keep.
- **KEY child-classifier** (`coerceMapKey`): unchanged from M2b — a key that does not cleanly match declines the whole map. **Live capture proved the dynamic bridge keeps non-matching enum / string-literal / literal-union keys leniently (full map, not partial)**, so a key miss is a deferred Mcoerce-d keep, never a native skip.
- **Decline preserved:** duplicate original keys (unproven insert/overwrite order) and non-object input.
- **Nullable clean-only:** map score is OWN flags only (`score.rs`) — a kept value's flags don't propagate — so a nullable optional map arm declines against the scored null arm. Mcoerce-c does not score collection arms vs null.

### Map child-classifier design (values + keys)
| child | proven-error → **SKIP entry** | deferred/uncertain → **DECLINE whole map** |
|---|---|---|
| **value** | numeric-string re-derivation (`bamlStringNumberFails`), `error_unexpected_type` kinds (object/bool/null → int; number/object/null → bool), certain match_string miss, multi-field all-required-flat-leaf class ← scalar | non-string → string/enum/literal (`JsonToString`/`ObjectToString`), object → literal (`ObjectToPrimitive`), object-missing-required class (default-fill), array (array-to-singular/M3), union scoring, case-fold uncertain |
| **key** | *(none — dynamic bridge keeps non-matching keys leniently)* | any non-clean match_string verdict: enum / string-literal / literal-union miss (full-map keep = Mcoerce-d), or case-fold uncertain |

### Corpus differential dispositions (all wants live-captured; gating diff green)
**Flipped to CLAIMED**
- `map_value_partial_bad_int` — `{"a":1,"bad":"nope","c":"7"}` → `{"a":1,"c":7}` (proven int-parse-error value skipped).
- `map_value_lenient_kept` — `{"a":"7","b":2.5}` → `{"a":7,"b":3}` (Mcoerce-b lenient values kept; map score ignores FloatToInt).
- `map_bad_value_type` (existing, false→**true**) — `{"a":1,"b":{"x":1}}` → `{"a":1}` (object value → int is a proven `error_unexpected_type` skip).

**Stay / pinned FALLBACK** (native declines; each justified)
- `map_literal_key_partial_bad_key`, `map_bad_key_original_order` — **live capture showed FULL maps**: the dynamic bridge keeps a non-matching `"A"|"B"` literal-union key, so a key miss is Mcoerce-d, not a skip.
- `map_enum_key_nonmember_live_probe` — non-member enum key kept (full map), same lenient-keep finding as `map_bad_enum_key`.
- `map_duplicate_key_stays_fallback` — duplicate original key (BAML insert/overwrite order unproven).
- `map_string_string_non_string_value_stays_fallback` — number value → `JsonToString` (Mcoerce-d).
- `nullable_optional_map_object_stays_fallback` — `ObjectToMap` is score-bearing (M3 vs null).
- `union_map_value_partial` — existing over-claim guard (partial map arm + string arm → pick_best).

### Validation
`gofmt` · `go build ./...` · `go vet ./...` + `-tags=integration` · `go test ./internal/debaml ./bamlutils ./worker ./dynclient/...` · `GOWORK=off go test ./codegen` (from `adapters/common`) · `go run ./cmd/verify-framework-adapter` (6/6) — all green. **bamlfuzz parse-recovery differential ran locally** (Docker + cached BAML v0.223.0 CFFI): native-vs-BAML diff green, every fixture's disposition matches its pin (60 claimed / 45 fallback).

Scope: `internal/debaml/{coerce,*_test}.go` + `integration/bamlfuzz_parse_recovery_test.go` + corpus JSON only. Runtime gate stays `BAML_REST_USE_DEBAML`; no seam/regen/embed/go.mod changes.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

<a href="https://cubic.dev/pr/invakid404/baml-rest/pull/561?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

## Summary by CodeRabbit

* **New Features**
  * Enhanced map parse-recovery now returns partial results when only specific map entries fail to coerce.
  * Preserves original input key order in recovered map outputs.
  * Improves lenient map value coercion (e.g., numeric strings and float-to-int rounding where applicable).

* **Bug Fixes**
  * Duplicate-key map inputs now reliably decline the entire map.
  * Non-clean map-key misses now decline the whole map, while proven value errors can be skipped without discarding valid entries.

* **Tests**
  * Expanded parse-recovery fixtures and added broader map coercion test coverage for edge cases.

### CODE REVIEW

File: internal/debaml/coerce_map_test.go [hunk 1]
```diff
@@ -0,0 +1,238 @@
...
+// TestCoerceMap_DeferredValuePreemptsKeySkip pins the other side of value-first:
+// a value that is a DEFERRED Mcoerce-d success (JsonToString) declines the WHOLE
+// map even when the key would otherwise be a proven skip — native must not emit
+// a partial map that drops an entry BAML would keep via a lenient value path.
+func TestCoerceMap_DeferredValuePreemptsKeySkip(t *testing.T) {
+	// map<"A"|"B", string> with {"C": 5}: key "C" would be a proven MapKeyParseError
+	// skip, but the value 5 -> string is JsonToString (Mcoerce-d, NOT proven), so
+	// the whole map declines rather than skipping the entry.
+	requireUnsupported(t, mapLitUnionStrSchema(), `{"u":{"C":5}}`)
+}
```
[the remark this answers is not in the document]
Comment (invakid404) [reply internal/debaml/coerce_map_test.go]: Fixed in `8880adb2` (Codex-verified). The `TestCoerceMap_DeferredValuePreemptsKeySkip` comment now accurately states `coerceMapKey` is decline-only — a non-matching literal-union key is a deferred lenient keep that declines the WHOLE map, not a `MapKeyParseError` partial skip. The assertion is unchanged (the value `5`→string is `JsonToString`/Mcoerce-d, so the map declines on the value before key `"C"` is even reached).

File: internal/debaml/coerce.go [hunk 2]
```diff
@@ -83,12 +83,16 @@ func (f *coerceFlags) isUncertain() bool { return f != nil && f.uncertain }
 // LIST parity (coerceList / coerceArray.rs): non-array SINGLE-TO-ARRAY wrapping,
 // PARTIAL array skips of PROVEN-parse-error items, and empty-list-on-singleton-
 // failure — each score-bearing (SingleToArray, ArrayItemParseError, child flags)
-// so the nullable-union clean-only rule still holds. The remaining leniencies stay
+// so the nullable-union clean-only rule still holds. Mcoerce-c also adds native
+// MAP parity (coerceMap / coerce_map.rs): object→map ObjectToMap flagging, VALUE-
+// then-KEY coercion, PARTIAL entry skips of PROVEN value/key parse errors
+// (MapValueParseError / MapKeyParseError), and ORIGINAL input key strings in
```
[the remark this answers is not in the document]
Comment (invakid404) [reply internal/debaml/coerce.go]: Fixed in `9c32fcb0` (Codex-verified). The `coerce` summary doc block previously said map value/key parse errors both partial-skip (`MapValueParseError`/`MapKeyParseError`); it now correctly states only **values** partial-skip on a proven `MapValueParseError`, while **keys decline the whole map** on any non-clean enum/literal/literal-union verdict (no `MapKeyParseError` skip), because the dynamic bridge keeps non-matching keys leniently (`try_cast_map` retains raw keys before `coerce_map` runs).

File: internal/debaml/coerce_map_test.go [hunk 3]
```diff
@@ -0,0 +1,238 @@
...
+	})
+	mustParseExact(t, s, `{"u":{"É":"zzz"}}`, `{"u":{}}`)
+}
+
+// TestCoerceMap_DeferredValuePreemptsKeySkip pins the other side of value-first:
+// a value that is a DEFERRED Mcoerce-d success (JsonToString) declines the WHOLE
+// map even when the key would otherwise be a proven skip — native must not emit
+// a partial map that drops an entry BAML would keep via a lenient value path.
+func TestCoerceMap_DeferredValuePreemptsKeySkip(t *testing.T) {
+	// map<"A"|"B", string> with {"C": 5}: key "C" would be a proven MapKeyParseError
```
[the remark this answers is not in the document]
Comment (invakid404) [reply internal/debaml/coerce_map_test.go]: Already addressed in `8880adb2` (duplicate of the CodeRabbit test-comment finding). The current `TestCoerceMap_DeferredValuePreemptsKeySkip` comment already conveys the value-first ordering you describe: the value `5`→string is `JsonToString` (Mcoerce-d, not a proven parse error) so the map declines on the value first — key `"C"` is never reached — and even if it were, `coerceMapKey` is decline-only (a whole-map decline, never a partial `MapKeyParseError` skip; see `TestCoerceMap_KeyMissDeclines`). Codex confirmed no further refinement needed. Resolving.
